File: incomplete_fizz_buzz/php/lib/FizzBuzz.php
<?php

class FizzBuzz
{

  public static function fizz_buzz($number, $printer)
  {
    if($number === 0) {
      $printer->execute(strval($number));
    } else if($number % 15 === 0) {
      $printer->execute('FizzBuzz');
    } else if($number % 3 === 0) {
      $printer->execute('Fizz');
    } else if($number % 5 === 0) {
      $printer->execute('Buzz');
    } else {
      $printer->execute(strval($number));
    }
  }

  public static function fizz_buzz_history($number, $printer)
  {
    if($number === 0) {
      $printer->execute("${number}, ${number}");
    } else if($number % 15 === 0) {
      $printer->execute("${number}, FizzBuzz");
    } else if($number % 3 === 0) {
      $printer->execute("${number}, Fizz");
    } else if($number % 5 === 0) {
      $printer->execute("${number}, Buzz");
    } else {
      $printer->execute("${number}, ${number}");
    }
  }

  public static function write($histories)
  {
    $file = fopen('data.txt', 'w');
    for($i = 0; $i < count($histories); $i++) {
      $history = $histories[$i];
      if($history === 0) {
        fwrite($file, "${history}, ${history}\n");
      } else if($history % 15 === 0) {
        fwrite($file, "${history}, FizzBuzz\n");
      } else if($history % 3 === 0) {
        fwrite($file, "${history}, Fizz\n");
      } else if($history % 5 === 0) {
        fwrite($file, "${history}, Buzz\n");
      } else {
        fwrite($file, "${history}, ${history}\n");
      }
    }
    fclose($file);
  }
}

File: incomplete_fizz_buzz/python/src/fizz_buzz.py
import os

def fizz_buzz(number, printer):
    if number == 0:
        printer.execute(str(number))
    elif number % 15 == 0:
        printer.execute('FizzBuzz')
    elif number % 3 == 0:
        printer.execute('Fizz')
    elif number % 5 == 0:
        printer.execute('Buzz')
    else:
        printer.execute(str(number))

def fizz_buzz_history(number, printer):
    if number == 0:
        printer.execute(str(number) + ', ' + str(number))
    elif number % 15 == 0:
        printer.execute(str(number) + ', FizzBuzz')
    elif number % 3 == 0:
        printer.execute(str(number) + ', Fizz')
    elif number % 5 == 0:
        printer.execute(str(number) + ', Buzz')
    else:
        printer.execute(str(number) + ', ' + str(number))

def write(histories, file_name):
    path = 'data/' + file_name
    file = open(path, 'w')
    for i in range(0, len(histories)):
        history = histories[i]
        if history == 0:
            file.write(str(history) + ', ' + str(history) + '\n')
        elif history % 15 == 0:
            file.write(str(history) + ', FizzBuzz' + '\n')
        elif history % 3 == 0:
            file.write(str(history) + ', Fizz' + '\n')
        elif history % 5 == 0:
            file.write(str(history) + ', Buzz' + '\n')
        else:
            file.write(str(history) + ', ' + str(history) + '\n')
    print('"%s" was saved.' % file_name)

def read(histories, printer, file_name):
    path = 'data/' + file_name
    if os.path.exists(path):
        file = open(path, 'r')
        while True:
            line = file.readline()
            if not line:
                break
            printer.execute(line.rstrip())
            file.close
    else:
        printer.error('"%s" is not found.' % file_name)
